elytron/WFCORE-4058-automatic-keystore-detection.adoc

```diff
@@ -0,0 +1,92 @@
+= Automatic detection of file based KeyStore types.
+:author:            Maritn Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+* If a user does not specify the type of a file based KeyStore can we auto detect the type?
+
+* There is most likely a finite set of types we would want to automatically recognise, possibly we could inspect the
+contents ourselves to detect different types - alternative attempt initialisation of different KeyStore objects until one succeeds. With the latter we would need to be careful one doesn't silently succeed when it should have failed.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/EAP7-675[EAP7-675]
+* https://issues.jboss.org/browse/ELY-1650[ELY-1650]
+* https://issues.jboss.org/browse/WFCORE-4058[WFCORE-4058]
+
+//=== Related Issues
+
+//* https://issues.jboss.org/browse/WFLY[WFLY-XXXX]
+
+=== Dev Contacts
+
+* mailto:mmazanek@redhat.com[Martin Mazanek]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+== Requirements
+
+=== Hard Requirements
+
+* Point of this change is to be able to load keystore without specifying its type.
+
+== Implementation Plan
+
+* Finite set of supported keystore types for automatic detection: JKS, JCEKS, PKCS12, BKS, BCFKS, UBER
+* Some types are fairly easily recognizable, as they start with magic number: JKS (0xfeedfeed), JCEKS (0xcececece) - although if a keystore starts with such number it doesn't necessarily mean that it is that type
+* Detect types of encrypted keystores? They may be a good candidate for try until success (let the provider do the decryption for us)
+
+* The idea is to prioritize different techniques with a priority
+** Magic number - check if the store contains one of known magic numbers - if yes, then try to load it and see if it is really that keystore
+** File extension - try to parse as a keystore based on the file extension
+** Try remaining supported types in a specific order / don't try anything and fail
```

Comment:
These are opposite ways. So which we will take? 

My concern is performance. You said it is intended to try only on server start (what about keystore reloads?) I assume biggest time consumption will be with trying to load keystores. How many keystores in worst case server will load? Could you approximate how many miliseconds it could take?

Reply:
Luckily I managed to make it 2 keystore loads as the worst case scenario, so performance shouldn't be a big problem. About time I can't guess, but the failed try will take less time than success, so in theory it should take less time than loading 2 keystores with defined type.

It will be implemented in the KeyStoreService, reaload should work fine and will use the recognized type - only one keystore load (or should I make it to recognize again, if someone decided to change the keystore format on running server?)

Reply:
I would say, changing keystore format on running server will be marginal use case. But when we say it will be automatic, I would say we should cover that scenario as well.

Can you measure duration of worst case detection? To know exactly what performance lost are we talking here about? So from what was said, I assume it will be just couple of milliseconds.

So again. You should choose in analysis form 
* Try remaining supported types in a specific order 
* Don't try anything and fail